docs/Tutorial_DiscordMERLIN.ipynb

```diff
@@ -0,0 +1,1356 @@
+{
```

Comment:
<pre class="ql-syntax" spellcheck="false">Line #68.            for n_attempts in range(1, max_attempts+1):
</pre><p>Instead of this, could we do:</p><p> </p><pre class="ql-syntax" spellcheck="false">if min_dist &lt;= 0:
    break

min_dists = min_dist * (1 - np.arange(1, int(np.ceil(1.0/decay[i]) + 1)) * decay[i])
min_dists = np.clip(min_dists, a_min=0.0, a_max=None)

for min_dist in min_dists:
    if np.all(exclude):
&nbsp; &nbsp; &nbsp; &nbsp; break
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;&nbsp;
&nbsp; &nbsp; is_cands = _select_candidates(subseqs, min_dist, exclude)
&nbsp; &nbsp; is_cands = _prune_candidates(subseqs, min_dist, is_cands)

    if np.any(is_cands):&nbsp;
&nbsp; &nbsp; &nbsp; &nbsp; break
</pre><p> </p><p>I think this may be a bit cleaner?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/TDAmeritrade/stumpy/pull/505/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='40'/>

Reply:
<p>Removing <code>if/else</code>  really makes the code cleaner! Thanks for your work. I have a few suggestions, please let me know if they sound reasonable from your perspective.</p><p> </p><ul><li>I think <code>np.arange(1, int(np.ceil(1.0/decay[i]) + 1))</code>  is a little bit long (it was hard for me to follow the parenthesis!). Is it okay to define a (temporary) variable <code>max_attemps=int(np.ceil(1.0/decay[i])</code>.Or, is that redundant?</li><li>Is it okay to use the plural <code>min_dists</code>  while we already have <code>min_dist</code> ? I mean, isn't it better to choose a more distinct name, say <code>min_dist_options</code>  or <code>min_dist_vec</code>? </li><li>With <code>min_dists = min_dist * (1 - np.arange(1, int(np.ceil(1.0/decay[i]) + 1)) * decay[i])</code>in the beginning, we are decreasing <code>min_dist</code>  even for the first attempt when <code>i=0</code>,  which is not necessary. We can add: </li></ul><p> </p><p><code>if i==0:</code></p><p><code>   min_dists = np.insert(min_dists,0,min_dist)</code></p>

Reply:
<blockquote>I think&nbsp;<code>np.arange(1, int(np.ceil(1.0/decay[i]) + 1))</code>&nbsp;is a little bit long (it was hard for me to follow the parenthesis!). Is it okay to define a (temporary) variable&nbsp;<code>max_attemps=int(np.ceil(1.0/decay[i])</code>.Or, is that redundant?</blockquote><p>Sounds good. Let's try it and see how it looks!</p><p> </p><blockquote>Is it okay to use the plural&nbsp;<code>min_dists</code>&nbsp;while we already have&nbsp;<code>min_dist</code>&nbsp;? I mean, isn't it better to choose a more distinct name, say&nbsp;<code>min_dist_options</code>&nbsp;or&nbsp;<code>min_dist_vec</code>?</blockquote><p>For visual distinctness, we can do <code>min_distances</code> ? Is that distinct enough from <code>min_dist</code> ?</p><p> </p><blockquote>With&nbsp;<code>min_dists = min_dist * (1 - np.arange(1, int(np.ceil(1.0/decay[i]) + 1)) * decay[i])</code>in the beginning, we are decreasing&nbsp;<code>min_dist</code>&nbsp;even for the first attempt when&nbsp;<code>i=0</code>, which is not necessary. We can add:</blockquote><p> </p><p> </p><p>Couldn't we resolve this by doing:</p><pre class="ql-syntax" spellcheck="false">max_attempts = int(np.ceil(1.0/decay[i]) + 1
min_distances = min_dist * (1 - np.arange(max_attempts)) * decay[i])
</pre><p>Note that I changed <code>np.arange</code>  to start at <code>0</code>  instead of starting from <code>1</code> .</p><p> </p><p>In fact, I always felt like <code>for n_attempts in range(1, max_attempts+1):</code>  felt funny to me. It's a lot better to start at zero and now the code is even cleaner I think.</p>

Reply:
> For visual distinctness, we can do `min_distances` ?

That works!

> Couldn't we resolve this by doing:
max_attempts = int(np.ceil(1.0/decay[i]) + 1
min_distances = min_dist * (1 - np.arange(max_attempts)) * decay[i])
Note that I changed np.arange to start at 0 instead of starting from 1 

Not really. Because, in each iteration, we would like to start with a `min_dist` that is lower than the previous iteration (except for the first attempt in `i=0`) This becomes important IF we decide to return as many as discords possible from `is_cands`. Although going with the same `min_dist` as previous iteration doesn't lead to wrong result, we are wasting time as we will end up with candidates that have been previously considered for the discords)

Reply:
<blockquote>This becomes important IF we decide to return as many as discords possible from&nbsp;<code>is_cands</code>.</blockquote><p>It's good to keep this in mind but let's worry about this later. </p><p> </p><blockquote>Although going with the same&nbsp;<code>min_dist</code>&nbsp;as previous iteration doesn't lead to wrong result, we are wasting time as we will end up with candidates that have been previously considered for the discords)</blockquote><p>Exactly! In fact, I think that even though "we will end up with candidates that have been previously considered":</p><p> </p><ol><li>You will exclude some candidates (in <code>exclude</code>)  and so there would be some less candidates (that have been assigned in the previous iteration). </li><li>When you decrease <code>min_dist</code>,  you will actually increase the number of candidates</li></ol><p> </p><p>So, I actually think that we will be fine and the "waste" isn't as bad as we think and, possibly, it might be good enough</p>

Reply:
Thanks for the feedback! Sounds good!